### ISSUE TITLE
System Map is not updated after hyperjump

### ISSUE THREAD

**Issue by Mc-Pain** [pioneerspacesim/pioneer#5691]
1) Plot a route
2) Jump to a next system
3) Open system overview

Observed: either departure system or target system is "shown" (Unexplored system. System view unavailable)

Expected: current system is shown

**Comment by Web-eWorks** [comment]
When opening the system overview, the *selected system* from the Sector Map is shown. Whichever system was selected before jumping to the next system in the list will be selected when existing hyperspace. We don't reset the user's selection in the Sector Map on jumping to a new system.

**Comment by Mc-Pain** [comment]
After playing E:D I would not expect that in the first place.

Suppose I am travelling far far away and eventually looking for a gas giant to refuel my ship (I am in uninhabited zone, so no spaceports available)
Route has been plotted, I'm running out of fuel, and after jump I am forced to
1) Go to the sector map
2) Click on "Reset View"
3) Go back to System Map

This is most unexpected to show target system contents unless explored.

Maybe we can track from where system map has been requested?
If sector map, show selected
Otherwise, show current

**Comment by Web-eWorks** [comment]
I'll re-open this for further discussion and let others chime in - in general I think the existing interaction pattern is well-defined and is less confusing/time-consuming than the alternatives, however I'm open to suggestions.

**Comment by nozmajner** [comment]
How about showing the system list of the route planner in the sys map and overview screens, so one can click through the route to see what to expect?
![route-overview](https://github.com/pioneerspacesim/pioneer/assets/4182678/c3e29ad2-3649-4051-af6e-1fea29c3793b)
![route-sysmap](https://github.com/pioneerspacesim/pioneer/assets/4182678/924c1eaf-38d4-4b88-82fa-317b234bf30b)

**Comment by mwerle** [comment]
Funnily enough, I worked on "fixing" this yesterday as I also find the current approach very unintuitive.  As I couldn't find an existing issue about this I opened a [thread on the forum](https://forum.pioneerspacesim.net/viewtopic.php?p=7259#p7259) to discuss this.

Here's a proposed patch:

https://github.com/pioneerspacesim/pioneer/compare/master...mwerle:pioneer:feat/reset_selected_system?expand=1

> "We don't reset the user's selection in the Sector Map on jumping to a new system."

Yes, obviously, hence this issue. But is that a historical oversight or the purposefully designed and intended behaviour? Apparently I'm not the only one that finds this behaviour extremely unintuitive..

What is the user-story behind the current behaviour? I fail to understand why I would want to see the details of a system which I'm not currently in, unless I've just selected it from the System Map.

The user-story behind wanting to see the current system after a jump is that I want to check it out - either I've arrived at my target system and want to set an in-system destination, or I'm just curious what systems I'm passing through (perhaps to scan anything interesting), or I'm part-way through a route and, as the OP mentioned, need to stop for resupplies (either from a station or scooping).

**Comment by Web-eWorks** [comment]
> What is the user-story behind the current behaviour? I fail to understand why I would want to see the details of a system which I'm not currently in, unless I've just selected it from the System Map.

I can see the existing behavior being desired when a user selects a system from the Sector Map to inspect in the system view, jumps, and then switches back to the system view while waiting for hyperspace travel to finish and the cooldown to complete.

However, I don't see a compelling reason in this case to keep the existing behavior as described over one that more naturally maps to a user's expectations.

> Here's a proposed patch:

Thanks for following up on the comment left in the code! I should point out that the behavior in your patch can be accomplished (in a less-invasive fashion) by adding the following code:

```lua
sectorView:ResetHyperspaceTarget()
sectorView:SwitchToPath(Game.system.path)
```

To `data/pigui/views/map-sector-view.lua` inside the `if ship:IsPlayer() then ...` block in the `onEnterSystem` event handler. Additionally, if you expose the `SectorView::SetSelected` method in `src/lua/LuaSectorView.cpp` and call that instead of `SwitchToPath()` the resulting behavior should be identical to the branch you've proposed.

While signals are a good tool to use, this is quite cosmetic in nature and handling it through the existing Lua signalling is perfectly fine instead of adding a new C++-side signal to Game's public API for a single one-off call.

**Comment by mwerle** [comment]
> I can see the existing behavior being desired when a user selects a system from the Sector Map to inspect in the system view, jumps, and then switches back to the system view while waiting for hyperspace travel to finish and the cooldown to complete.

This still happens, although on arrival in the new system, the map will change (if the new system is not the previously-selected system), which might be a bit unexpected.

----

I'm a C programmer; I am afraid the entire interaction with Lua is something very foreign to me. If the intent of the comment was to use LUA instead, as you suggest, then it would be very good for the comment to have actually said so..  I see no signalling anywhere there - you call it "Lua signalling" but to me it appears to be simply invoking a Lua callback which, in turn, invokes the C++ functionality.

As-is, I investigated the rest of the code and figured out how to use the signalling library used by Pioneer; again, something I was not previously familiar with. 

What is the benefit of calling into the Lua handler only for it to call back into the C code as per your proposal? It seems to me to be significantly less performant... but if there is a benefit / programming pattern which I am missing I would like to learn about it.

**Comment by Web-eWorks** [comment]
> I'm a C programmer; I am afraid the entire interaction with Lua is something very foreign to me. If the intent of the comment was to use LUA instead, as you suggest, then it would be very good for the comment to have actually said so.. I see no signalling anywhere there - you call it "Lua signalling" but to me it appears to be simply invoking a Lua callback which, in turn, invokes the C++ functionality.

Right - for clarity, if you're interested in PR'ing a change I'd strongly prefer it be implemented in Lua as described above unless you can provide a good reason to do it in C++ - for example, if you had future PRs planned which would also use the proposed signal to clean up problem areas of the code.

> As-is, I investigated the rest of the code and figured out how to use the signalling library used by Pioneer; again, something I was not previously familiar with.

I very much appreciate that! In this particular case the comment left in the code is not / no longer the "optimal" solution, which is why I was suggesting the change if you are willing to adopt and PR it.

> What is the benefit of calling into the Lua handler only for it to call back into the C code as per your proposal? It seems to me to be significantly less performant... but if there is a benefit / programming pattern which I am missing I would like to learn about it.

The main reason for this is that it can be easily extended to disable the behavior based on player intent (i.e. if the player has explicitly clicked a hypothetical "View This System in System Map" button), does not involve changes to the "game framework" C++ API, and can be edited by the greatest number of potential contributors.

While Lua is less performant than compiled C++, this is not an area where performance is anything approaching a concern. It's a one-off event that happens once every few seconds of real time at most, and the overhead of calling back and forth with Lua is on the order of a few microseconds at worst. Additionally, the behavior is then "located" with other aspects of the sector map which respond to game events.

The primary and most fundamental benefit however is that the behavior can be changed by future contributors, modders, etc. who don't know C++ or the signal library we use, and do not need to recompile the game to see changes. Lua is significantly easier to learn than C++, and the codebase tries to follow a principle of approachability over maximum theoretical performance.

**Comment by impaktor** [comment]
Changing this current default behavior is :+1:  from my side.

(If the change is in Lua, some future modder can add a checkbox "[ ] selection follows post jump location")

**Comment by Gliese852** [comment]
We can remember how it was in Frontier - system map *always* showed the current system. For a selected system, you can get into the overview screen, and only from there you can get into the 3D map of the selected system. That is, when jumping, the selected system does not change, which is nice. There was a similar proposal from @Mc-Pain:

> Maybe we can track from where system map has been requested?
> If sector map, show selected
> Otherwise, show current

**Comment by mwerle** [comment]
I can certainly work on my code to add more features; ie, track from where the screens were opened to see whether it should show the current system or selected system. Or even add a checkbox "selection follows jump" (although that might be a stretch goal as I don't do GUIs..) etc.

> Right - for clarity, if you're interested in PR'ing a change I'd strongly prefer it be implemented in Lua as described above unless you can provide a good reason to do it in C++ 

I have no good reason to do it in C++ save that that is what I'm familiar with. Thank you for your detailed reply - I always endeavour to follow a projects' preferred approach, but, being brand-new to this codebase, it will take me some time to work out what that is.

One potential issue with tracking from where the System Map / System Overview gets opened up from is that there are no function keys assigned to open those directly. "F2" opens the last map-view and from there you can use F5/F6/F7 to switch between the various maps (FWIW, F8 is documented as showing the Galaxy map, but that appears to do nothing on current builds).  Given most (all?) modern keyboards have 12 functions keys I do find it odd that the game restricts itself to using only 8 and then overloads a bunch of stuff. For example, "F5" is used to both launch as well as change flight modes, when there is already another key (I think it's "N" by default) set up for that, and, in map-mode, switches to Sector Map. 

It's been a while, but I seem to remember Frontier was a bit more consistent with the way its function-keys were used. It was also nice that the icons showed the key number as a reminder..

----

TL;DR: I'll rework my code to use Lua, and see if I can come up with a nice way to NOT change the screen if it's already open during a hyperspace jump.

**Comment by impaktor** [comment]
> (FWIW, F8 is documented as showing the Galaxy map, but that appears to do nothing on current builds)

It did not survive the migration to [PiGUI](https://wiki.pioneerspacesim.net/wiki/GUI_introduction#Background_history)

> It's been a while, but I seem to remember Frontier was a bit more consistent with the way its function-keys were used

I believe we follow their layout pretty much. However, we're not married to them, and ideally that whole layout structure should be re-organized to something more intuitive. Not sure what/how, though.

> It was also nice that the icons showed the key number as a reminder..

Downside: if player re-maps the keys, icons will now show wrong key.

**Comment by mwerle** [comment]
> Downside: if player re-maps the keys, icons will now show wrong key.
Can the "F" keys be remapped?  But yes, good point..

> ideally that whole layout structure should be re-organized to something more intuitive
+1 to this. To be honest, I find the UI a bit unintuitive, coming from long years of playing Elite and Elite:Dangerous (I did play Frontier and FFE a bit, but nowhere near as extensively).  Especially setting various flight modes and engaging the autopilot etc.  Needing to use both keyboard and mouse at the same time is, IMHO, suboptimal. It would be nice to be able to completely control the game by keyboard.  Or perhaps I'm missing something, I've only re-started playing Pioneer a week or so ago.

</offtopic>

---

I've changed the code to move the logic to Lua, as suggested.

Given that, broadly speaking, there is support for this change, I have now created a draft PR - #5922 

I guess the detailed discussion about the implementation should continue there?

**Comment by Web-eWorks** [comment]
> Can the "F" keys be remapped? But yes, good point..

They "should" be able to be remapped for completeness. This is not currently possible however, as they're using an interim input subsystem with the planned intent to migrate them to the same input system the rest of the game code uses. That migration is blocked on exposing the primary input subsystem to Lua, which is a low-priority task falling on my shoulders that keeps getting delayed by other tasks and has its tendrils tangled in mod support and other low-level concerns.

> It would be nice to be able to completely control the flight-mode part of the game by keyboard or HOTAS. Or perhaps I'm missing something, I've only re-started playing Pioneer a week or so ago.

The flight mode radials specifically have support for keyboard / HOTAS bindings - I forget the specific key binding names off-hand but there are a set of bindings for general radial direction selection and two bindings that open the specific radial menus.

> I've changed the code to move the logic to Lua, as suggested.

Thank you! I'll be quite happy to give that a look-over and suggest changes or approve sometime tomorrow (in +12 hours as of this writing).

**Comment by mwerle** [comment]
> The flight mode radials specifically have support for keyboard / HOTAS bindings

Thank you. I found 2, but not sure they handled all things.. again, I'm new to the current version of Pioneer so definitely don't know everything about the game yet!  That said, the radial popups didn't have a key assigned by default (AppImage build), nor are they mentioned in the user manuals (at least, not in the bits I've looked at so far).

> Thank you! I'll be quite happy to give that a look-over and suggest changes or approve sometime tomorrow (in +12 hours as of this writing).

No rush; I'm off scuba-diving this weekend (although it's looking rather miserable, so might take a laptop along..)

**Comment by impaktor** [comment]
> the radial popups didn't have a key assigned by default (AppImage build), nor are they mentioned in the user manuals (at least, not in the bits I've looked at so far).

Correct. There's never been any keybindings to the radial menus. ...until soon? (hint hint, wink wink)
